contract/src/agoric-basics.contract.js

```diff
@@ -40,58 +40,112 @@ const bagCounts = bag => {
   const entries = getCopyBagEntries(bag);
   return entries.map(([_k, ct]) => ct);
 };
+
+/**
+ *
+ * @param {import('@endo/patterns').CopyBag} bag
+ * @param {Object.<string, {tradePrice: Amount, maxTickets: bigint}>} inventory
+ * @returns {boolean}
+ */
+export const hasInventory = (bag, inventory) => {
+  const entries = getCopyBagEntries(bag);
+  for (const [k, ct] of entries) {
+    if (k in inventory) {
+      const { maxTickets } = inventory[k];
+      if (ct > maxTickets) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+/**
+ *
+ * @param {Amount} amount
+ * @param {number} n
+ * @returns {Amount}
+ */
+const multiply = (amount, n) => {
+  const arr = Array.from({ length: n });
+  return arr.reduce(
+    (sum, _) => AmountMath.add(amount, sum),
+    AmountMath.make(amount.brand, 0n),
+  );
+};
+
+/**
+ *
+ * @param {Amount} sum
+ * @param {[string, bigint]} entry
+ * @param {Object.<string, {tradePrice: Amount, maxTickets: bigint}>} inventory
+ * @returns {Amount}
+ */
+const addMultiples = (sum, entry, inventory) => {
+  const multiple = multiply(inventory[entry[0]].tradePrice, Number(entry[1]));
+  return AmountMath.add(multiple, sum);
+};
+
+/**
+ *
+ * @param {import('@endo/patterns').CopyBag} bag
+ * @param {Object.<string, {tradePrice: Amount, maxTickets: bigint}>} inventory
+ * @returns {Amount}
+ */
+export const bagPrice = (bag, inventory) => {
+  const entries = getCopyBagEntries(bag);
+  const brand = Object.values(inventory)[0].tradePrice.brand;
+  return entries.reduce(
+    (sum, entry) => addMultiples(sum, entry, inventory),
```

Comment:
_question_

Not sure what's the best way to add JSDoc type annotations for anonymous functions like this?

Reply:
In this case, you shouldn't have to.

The problem is that the return type of `getCopyBagEntries` is `any` due a bug related to https://github.com/endojs/endo/issues/1488

If we add a declaration to work around that, the type of this anonymous function is inferred.

```diff
@@ -84,6 +84,7 @@ const addMultiples = (sum, entry, inventory) => {
  * @returns {Amount}
  */
 export const bagPrice = (bag, inventory) => {
+  /** @type {[string, bigint][]} */
   const entries = getCopyBagEntries(bag);
```